### DISCUSSION TITLE
RPC anonymous users access to `auth` schema

### DISCUSSION THREAD

**Discussion by vabatta** [supabase/supabase#5460] [Questions]
Hi everyone,

I have a `carts` table that allows authenticated and anonymous users to have a personal cart. If a cart belongs to an authenticated user, only that user can access it. 
The function `retrieve_cart` either retrieve a cart by its ID (if accessible due to policies) or create a new one. The problem is that when creating a new one, I want it to be directly associated with the user creator (if any) using `auth.uid()` but RPC gives the error `permission denied for schema auth`. I'm wondering which implications would occur if I actually do `grant usage to auth to anon`.

```sql
create table carts (
  id uuid default uuid_generate_v4() not null,
  user_id uuid,
  primary key (id),
  foreign key (user_id) references auth.users(id) on delete cascade
);

create policy "Enable access to all users for anonymous or personal carts" on carts for 
  select using (user_id is null or user_id = auth.uid());

create policy "Enable insertion to all users for anonymous or personal carts" on carts for 
  insert with check (user_id is null or user_id = auth.uid());

create or replace function retrieve_cart(cart_id uuid default uuid_generate_v4()) returns carts as $$
  with new_or_existing as (
    -- removing auth.uid() makes the function working perfectly, but an authenticated users will not have its cart automatically assigned on creation
    insert into carts (id, user_id) values (retrieve_cart.cart_id, auth.uid()) 
    on conflict (id) do nothing
    returning *
  )
  select * from new_or_existing
  union
  select * from carts where id = retrieve_cart.cart_id;
$$ language sql strict volatile security invoker;
```

Any idea on how I could improve the above without security implication or pesky workarounds (eg. triggers on insert)?

**Comment by GaryAustin1** [comment]
You want "security definer" in your function to allow it to act with privileges of the creator of the function. 
But you then need to make sure you handle any type of security that was done by RLS in the function itself.

**Reply by vabatta** [reply]
So there is no way to use `security invoker` and achieve the above request? I don't like the idea of making function `security definer` because over time it's easy to lose track of policies or accesses changes.

**Reply by GaryAustin1** [reply]
One thing bothering me, but I don't have time to look it up today.  I did not think auth.uid() was a protected function.  You can use it as default in creating a column in a new row....  I may be missing something or because in that case user does not actually have access, like RLS it is OK.
You can also get the uid from the jwt config value just like the auth.uid() function does...

**Reply by vabatta** [reply]
I've solved by using these two grants (but I'm not sure about the security implications):
```sql
grant execute on all functions in schema auth to anon;
grant usage on schema auth to anon;
```
